src/lib.rs

```diff
@@ -48,63 +48,90 @@ pub struct rustls_client_session {
 const RUSTLS_CRATE_VERSION: &str = "0.19.0";
 
 #[macro_export]
-macro_rules! ffi_panic_boundary {
-    ( $($tt:tt)* ) => {
+macro_rules! ffi_panic_boundary_generic {
+    ( $retval:expr, $($tt:tt)* ) => {
         match ::std::panic::catch_unwind(|| {
             $($tt)*
         }) {
             Ok(ret) => ret,
-            Err(_) => return rustls_result::Panic,
+            Err(_) => return $retval,
         }
-  }
+    }
 }
 
 #[macro_export]
-macro_rules! ffi_panic_boundary_size_t {
+macro_rules! ffi_panic_boundary {
     ( $($tt:tt)* ) => {
         match ::std::panic::catch_unwind(|| {
             $($tt)*
         }) {
-            | Ok(ret) => ret,
-            | Err(_) => return 0,
+            Ok(ret) => ret,
+            Err(_) => return rustls_result::Panic,
         }
   }
 }
 
+#[macro_export]
+macro_rules! ffi_panic_boundary_size_t {
+    ( $($tt:tt)* ) => {
+        ffi_panic_boundary_generic!(0, $($tt)*)
+    }
+}
+
 #[macro_export]
 macro_rules! ffi_panic_boundary_bool {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return false,
-        }
-  }
+        ffi_panic_boundary_generic!(false, $($tt)*)
+    }
 }
 
 #[macro_export]
 macro_rules! ffi_panic_boundary_ptr {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return std::ptr::null_mut(),
-        }
-  }
+        ffi_panic_boundary_generic!(std::ptr::null_mut(), $($tt)*)
+    }
 }
 
 #[macro_export]
 macro_rules! ffi_panic_boundary_unit {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return,
-        }
-  }
+        ffi_panic_boundary_generic!((), $($tt)*)
+    }
+}
+
+/// If the provided pointer is non-null, convert it to a reference
+/// to the type in the second argument.
+/// Otherwise, return NullParameter (in the two-argument form) or the provided
+/// value (in the three-argument form).
+#[macro_export]
+macro_rules! checked_ptr_ref {
```

Comment:
Galaxy brain review comment: could you write a single macro that matches on the literal `mut` in front of the `$typ:ty`?

```
macro_rules! checked_ptr {
    ( $var:ident, mut $typ:ty, $retval:expr ) => {
        // handle mutable case with .as_mut()
    };
    ( $var:ident, $typ:ty, $retval:expr ) => {
       // handle immutable case with .as_ref()
    }
}
```

Or maybe there's some other way to write arms that match on either `mut T` vs `T`? Then you wouldn't have to define two nearly identical macros for mut vs. ref. This is definitely code golf, and I'm more curious about whether it's possible than advocating for it as a good idea.

Reply:
This was a cool idea and it worked! I didn't wind up de-duplicating the macro definitions (probably could but I felt like it would be too fancy). But by adding `&` and `&mut` to the matches, the call sites for the macro more clearly express the type being cast to.